src/boot/efi/util.c

```diff
@@ -769,6 +770,126 @@ EFI_STATUS make_file_device_path(EFI_HANDLE device, const char16_t *file, EFI_DE
         return EFI_SUCCESS;
 }
 
+static const EFI_GUID QEMU_KERNEL_LOADER_FS_MEDIA_GUID = {
+        0x1428f772, 0xb64a, 0x441e, {0xb8, 0xc3, 0x9e, 0xbd, 0xd7, 0xf8, 0x93, 0xc7 }};
+
+static const EFI_GUID QEMU_BOOT_ORDER_GUID = {
+        0x668f4529, 0x63d0, 0x4bb5, {0xb6, 0x5d, 0x6f, 0xbb, 0x9d, 0x36, 0xa4, 0x4a }};
+
+/* detect direct kernel boot, i.e. 'qemu -kernel systemd-bootx64.efi' */
+BOOLEAN is_qemu_direct_kernel(EFI_HANDLE device) {
+        EFI_STATUS err;
+        VENDOR_DEVICE_PATH *dp;
+
+        err = BS->HandleProtocol(device, &DevicePathProtocol, (void **) &dp);
+        if (err != EFI_SUCCESS)
+                return FALSE;
+
+        if (dp->Header.Type != MEDIA_DEVICE_PATH ||
+            dp->Header.SubType != MEDIA_VENDOR_DP)
+                return FALSE;
+
+        if (memcmp(&dp->Guid, &QEMU_KERNEL_LOADER_FS_MEDIA_GUID, sizeof(EFI_GUID)) != 0)
+                return FALSE;
+
+        Print(L"loaded via qemu direct kernel boot\n");
+        return TRUE;
+}
+
+/* detect loading from firmware volume (sd-boot added to ovmf) */
+BOOLEAN is_firmware_volume(EFI_HANDLE device) {
+        EFI_STATUS err;
+        VENDOR_DEVICE_PATH *dp;
+
+        err = BS->HandleProtocol(device, &DevicePathProtocol, (void **) &dp);
+        if (err != EFI_SUCCESS)
+                return FALSE;
+
+        if (dp->Header.Type != MEDIA_DEVICE_PATH ||
+            dp->Header.SubType != MEDIA_PIWG_FW_VOL_DP)
+                return FALSE;
+
+        Print(L"loaded from firmware volume\n");
+        return TRUE;
+}
+
+/* try find ESP when not loaded from ESP */
+EFI_STATUS qemu_open(EFI_HANDLE *qemu_dev, EFI_FILE **qemu_dir) {
+        _cleanup_free_ EFI_HANDLE *handles = NULL;
+        _cleanup_free_ CHAR16 *dp_str = NULL;
+        _cleanup_free_ CHAR16 *fs_str = NULL;
+        _cleanup_(file_closep) EFI_FILE *root_dir = NULL;
+        _cleanup_(file_closep) EFI_FILE *efi_dir = NULL;
+        EFI_DEVICE_PATH *dp = NULL;
+        EFI_DEVICE_PATH *fs = NULL;
+        CHAR16 order_str[20];
+        UINTN n_handles = 0;
+        UINTN order = 0;
+        EFI_STATUS err;
+        uint32_t item, dp_len;
+
+        // ovmf doesn't connect devices for direct kernel boot
+        reconnect();
+
+        // find all file system handles
+        err = BS->LocateHandleBuffer(ByProtocol, &FileSystemProtocol, NULL, &n_handles, &handles);
+        if (err != EFI_SUCCESS)
+                return err;
+
+        do {
+                free(dp_str); dp_str = NULL;
```

Comment:
Just move dp_str definition into the loop. Make it a `for(;;)` loop and have a `if(!dp_str) break;` at the end.